plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -1546,17 +1551,48 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni DefaultFun where
 
     toBuiltinMeaning _semvar NullList =
         let nullListDenotation :: SomeConstant uni [a] -> BuiltinResult Bool
-            nullListDenotation (SomeConstant (Some (ValueOf uniListA xs))) = do
+            nullListDenotation (SomeConstant (Some (ValueOf uniListA xs))) =
                 case uniListA of
-                    DefaultUniList _ -> pure $ null xs
-                    _                ->
-                        -- See Note [Structural vs operational errors within builtins].
-                        throwing _StructuralUnliftingError "Expected a list but got something else"
+                    DefaultUniList _argUni -> pure $ null xs
+                    _ -> throwing _StructuralUnliftingError "Expected a list but got something else"
             {-# INLINE nullListDenotation #-}
         in makeBuiltinMeaning
             nullListDenotation
             (runCostingFunOneArgument . paramNullList)
 
+    toBuiltinMeaning _semvar LengthArray =
+      let lengthArrayDenotation :: SomeConstant uni (Vector a) -> BuiltinResult Int
+          lengthArrayDenotation (SomeConstant (Some (ValueOf uni vec))) =
+            case uni of
+              DefaultUniArray _argUni -> pure $ Vector.length vec
+              _ -> throwing _StructuralUnliftingError "Expected an array but got something else"
+          {-# INLINE lengthArrayDenotation #-}
+        in makeBuiltinMeaning lengthArrayDenotation (runCostingFunOneArgument . paramLengthArray)
```

Comment:
I think this is what happens if you try to group related builtins together.  I've always been in the habit of adding new builtins at the very end of the datatype and adding code for dealing with them as the final clause of any function that has to be modified (although I see that this convention's already been violated for `caseData` and `caseList`), and that's why the comment about `unimplementedCostingFunction` is at the end, where I thought it would be visible when someone's adding a new builtin.   The flat tags certainly have to go at the end, and it seems reasonable to try and preserve the historical order that the functions have been added in elsewhere. There's also the tricky business of tagging the cost model parameter names that we have to do in `plutus-ledger-api` ([here](https://github.com/IntersectMBO/plutus/blob/e9b258319b09c2fa657b56841db6ab1cb9db4c6d/plutus-ledger-api/src/PlutusLedgerApi/V3/ParamName.hs#L13) for example) and it might be easier to keep track of that if we keep stuff in the same order everywhere.

Reply:
I wouldn't mind it we said that new builtins must go last and move `caseList` and `caseData` there.

Reply:
Mind you, in the version in master, `toBuiltinMeaning` and its documentation already take up 1804 lines, so it's hard to see where you could put a comment so that no-one would miss it.  The not-very-easy-to-find [document](https://github.com/IntersectMBO/plutus/blob/master/plutus-core/cost-model/CostModelGeneration.md#adding-a-new-function) about adding a new function and costing it also tell you to use `unimplementedCostingFunction`.